src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,429 @@
+ *     GNU Lesser General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Lesser General Public License
+ *     along with Discord4J.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package sx.blah.discord.util;
+
+import sx.blah.discord.handle.impl.obj.Channel;
+import sx.blah.discord.handle.obj.IChannel;
+import sx.blah.discord.handle.obj.IMessage;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static sx.blah.discord.handle.impl.obj.Channel.MESSAGE_CHUNK_COUNT;
+
+/**
+ * This is a utility class for building message histories in a readable, easy-to-understand fashion. It can be built
+ * into a {@link MessageHistory} object, a {@link Stream} of {@link IMessage} objects, or an {@link Iterator} of
+ * {@link IMessage} objects. It can also be directly used in a for-each loop without being built.
+ *
+ * @see MessageHistory
+ * @see MessageHistoryIterator
+ */
+public class MessageHistoryBuilder implements Iterable<IMessage> {
+
+	private Channel channel;
+
+	// User provided info
+	private IMessage uStart; // Null if not given
+	private LocalDateTime uStartTime; // null signifies most recent
+	private boolean uIncludeStart;
+	private IMessage uEnd; // Null if not given
+	private LocalDateTime uEndTime; // null signifies channel start
+	private boolean uIncludeEnd;
+	private int count = -1; // < 0 signifies no limit
+
+	// Used for builds
+	private List<IMessage> cached;
+	private List<IMessage> messages;
+	private MessageHistoryRange range;
+	private int currentCount;
+
+	public MessageHistoryBuilder(IChannel channel) {
+		this.channel = (Channel) channel;
+	}
+
+	/**
+	 * Specifies the message to start at (inclusive).
+	 *
+	 * @param startID the message ID to start at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder startAt(long startID) {
+		this.uStart = channel.getMessageByID(startID);
+		if (uStart == null)
+			throw new IllegalArgumentException("Message " + Long.toUnsignedString(startID) + " does not exist in channel " + channel.getStringID() + "!");
+		this.uStartTime = uStart.getTimestamp();
+		this.uIncludeStart = true;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to start at (inclusive).
+	 *
+	 * @param msg the message to start at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder startAt(IMessage msg) {
+		if (!channel.equals(msg.getChannel()))
+			throw new IllegalArgumentException("Message " + msg.getStringID() + " does not exist in channel " + channel.getStringID() + "!");
+		this.uStart = msg;
+		this.uStartTime = uStart.getTimestamp();
+		this.uIncludeStart = true;
+		return this;
+	}
+
+	/**
+	 * Specifies the time to start at (inclusive).
+	 * NOTE: Using this method to set a start time chronologically before endAt, and using withMaxCount > 0, may
```

Comment:
The only reason I provided a stream method is because it gets the messages in chunks. It isn't equal to #get().stream() because the latter fetches the entire range before streaming while the former only has fetched at most two chunks at once. 

It is a feature of the API. Message history can be traversed forwards or backwards (using the `before` and `after` params) and I wanted to reflect that in a convenient way. It does add a lot of complexity to the builder, but from the user's standpoint it is much easier.